File: webapp/services/fit_checker.py
import sys
from pathlib import Path

PROJECT_ROOT = Path(__file__).parent.parent.parent.resolve()
sys.path.insert(0, str(PROJECT_ROOT / 'tools'))

from glyph_map import encode_game_utf16, encode_game_sjis


def check_fit(translated_text: str, source: str, capacity: int,
              glyph_map: dict | None = None) -> dict:
    if not translated_text.strip():
        return {
            'status': 'unchecked',
            'used_bytes': 0,
            'capacity': capacity,
            'remaining': capacity,
        }

    if source == 'SCRIPT':
        encoded = encode_game_utf16(translated_text, glyph_map)
    else:
        encoded = encode_game_sjis(translated_text, glyph_map)

    used = len(encoded) + (2 if source == 'SCRIPT' else 0)
    remaining = capacity - used

    if remaining >= 20:
        status = 'ok'
    elif remaining >= 0:
        status = 'tight'
    else:
        status = 'needs_shift'

    return {
        'status': status,
        'used_bytes': used,
        'capacity': capacity,
        'remaining': remaining,
    }


def batch_check_fit(entries: list, session, glyph_map: dict | None = None) -> int:
    from ..database import TextEntry
    changed = 0
    for entry in entries:
        if not entry.translated_text:
            continue
        result = check_fit(
            entry.translated_text,
            entry.source,
            entry.segment_capacity or entry.original_bytes or 999,
            glyph_map=glyph_map,
        )
        if result['status'] != entry.fit_status:
            entry.fit_status = result['status']
            entry.needs_shift = (result['status'] == 'needs_shift')
            changed += 1
    if changed:
        session.commit()
    return changed

File: crates/straw-core/src/fit_checker.rs
use crate::glyph_map::{encode_game_sjis, encode_game_utf16, GlyphMap};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TextSource {
    Script,
    Elf,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum FitStatus {
    Unchecked,
    Ok,
    Tight,
    NeedsShift,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FitResult {
    pub status: FitStatus,
    pub used_bytes: usize,
    pub capacity: usize,
    pub remaining: isize,
}

pub fn check_fit(
    translated_text: &str,
    source: TextSource,
    capacity: usize,
    glyph_map: Option<&GlyphMap>,
) -> FitResult {
    if translated_text.trim().is_empty() {
        return FitResult {
            status: FitStatus::Unchecked,
            used_bytes: 0,
            capacity,
            remaining: capacity as isize,
        };
    }

    let encoded = match source {
        TextSource::Script => encode_game_utf16(translated_text, glyph_map),
        TextSource::Elf => encode_game_sjis(translated_text, glyph_map),
    };
    let terminator = if source == TextSource::Script { 2 } else { 0 };
    let used_bytes = encoded.len() + terminator;
    let remaining = capacity as isize - used_bytes as isize;
    let status = if remaining >= 20 {
        FitStatus::Ok
    } else if remaining >= 0 {
        FitStatus::Tight
    } else {
        FitStatus::NeedsShift
    };

    FitResult {
        status,
        used_bytes,
        capacity,
        remaining,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::glyph_map::{english_glyph_map, spanish_glyph_map};

    #[test]
    fn empty_text_is_unchecked() {
        let result = check_fit("", TextSource::Script, 100, None);
        assert_eq!(result.status, FitStatus::Unchecked);
        assert_eq!(result.remaining, 100);
    }

    #[test]
    fn script_counts_utf16_and_terminator() {
        let map = spanish_glyph_map();
        let result = check_fit("Hola", TextSource::Script, 100, Some(&map));
        assert_eq!(result.status, FitStatus::Ok);
        assert_eq!(result.used_bytes, 10);
    }

    #[test]
    fn detects_tight_and_needs_shift() {
        let map = english_glyph_map();
        assert_eq!(
            check_fit("x".repeat(40).as_str(), TextSource::Script, 82, Some(&map)).status,
            FitStatus::Tight
        );
        assert_eq!(
            check_fit("x".repeat(41).as_str(), TextSource::Script, 82, Some(&map)).status,
            FitStatus::NeedsShift
        );
    }
}
